sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableDestinationCoder.java

```diff
@@ -45,13 +46,20 @@ public void encode(TableDestination value, OutputStream outStream)
     }
     tableSpecCoder.encode(value.getTableSpec(), outStream);
     tableDescriptionCoder.encode(value.getTableDescription(), outStream);
+    timePartitioningCoder.encode(value.getJsonTimePartitioning(), outStream);
   }
 
   @Override
   public TableDestination decode(InputStream inStream) throws IOException {
     String tableSpec = tableSpecCoder.decode(inStream);
     String tableDescription = tableDescriptionCoder.decode(inStream);
-    return new TableDestination(tableSpec, tableDescription);
+    String jsonTimePartitioning = null;
+    try {
+      jsonTimePartitioning = timePartitioningCoder.decode(inStream);
+    } catch (IOException e) {
+      // This implies we're decoding old state that did not contain TimePartitioning. Continue.
```

Comment:
EOFException is not part of the contract of Coder in general, nor StringUtf8Coder in particular, and I don't think it's a good idea to make it part of the contract (and I don't know how to formalize it, even if we decided to make it part of the contract).

E.g. imagine that the current coder is wrapped into another coder, so there's actually some more data in the input stream, and decoding will not fail with an EOFException but will instead steal some data from the next encoded thing.

The approach of "if your data format is actually different, use a different coder" seems more sustainable to me. We likely want to eventually come up with a solution for migrating old data to new data during pipeline update, but I think catching EOFExceptions is not that solution.